scalafix-nsc/src/main/scala/scalafix/nsc/NscSemanticApi.scala

```diff
@@ -11,104 +11,402 @@ import scalafix.ScalafixConfig
+              case r => List(r)
+            }
+          case Nil => List(ref)
+        }
+      }
+
+      val transformedRef = ref.transform {
+        case m.Type.Select(qual, typeName) =>
+          val removed = loop(qual, prefixes.reverse)
+          if (removed.isEmpty) typeName
+          else m.Type.Select(removed.head, typeName)
+        case r => r
+      }
+
+      transformedRef.asInstanceOf[m.Ref]
+    }
+
+    /** Missing import and shortened name */
+    type Missing = (m.Ref, m.Ref)
+
+    /** Shortened name */
+    type Hit = m.Ref
+
+    /** Get the shortened type `ref` at a concrete spot. */
+    def getMissingOrHit(ref: m.Ref,
+                        inScope: g.Scope,
+                        enclosingTerm: g.Symbol): Either[Missing, Hit] = {
+
+      val refNoThis = stripThis(ref).asInstanceOf[m.Ref]
+      val names = refNoThis.collect {
+        case tn: m.Term.Name => tn
+        case tn: m.Type.Name => tn
+      }
+
+      // Mix local scope with root scope for FQN and non-FQN lookups
+      val wholeScope = mixScopes(inScope, realRootScope)
+      val disambiguatingSyntax = ref.syntax
+      val (_, reversedSymbols) = {
+        names.iterator.foldLeft(wholeScope -> List.empty[g.Symbol]) {
+          case ((scope, symbols), metaName) =>
+            val sym = lookupBothNames(metaName.value,
+                                      scope,
+                                      symbols.headOption,
+                                      disambiguatingSyntax)
+            if (!sym.exists) scope -> symbols
+            else sym.info.members -> (sym :: symbols)
+        }
+      }
+
+      val symbols = reversedSymbols.reverse
+      val metaToSymbols = names.zip(symbols)
+      logger.elem(metaToSymbols)
+
+      if (symbols.nonEmpty) {
+        /* Check for path dependent types:
+         * 1. Locate the term among the FQN
+         * 2. If it exists, get the first value in the chain */
+        val maybePathDependentType = metaToSymbols
+          .find(ms => ms._2.isAccessor)
+          .flatMap(_ =>
+            metaToSymbols.dropWhile(ms => !ms._2.isValue).headOption)
+        val isPathDependent = maybePathDependentType.isDefined
+
+        val (lastName, lastSymbol) = maybePathDependentType
+          .getOrElse(metaToSymbols.last)
+        val (onlyPaths, shortenedNames) =
+          metaToSymbols.span(_._1 != lastName)
+
+        val renames: Map[m.Name, g.Name] = shortenedNames.map {
+          case (metaName, symbol) =>
+            metaName -> st.renames.getOrElse(symbol, symbol).name
+        }.toMap
+
+        val localSym = inScope.lookup(lastSymbol.name)
+        if (lastSymbol.exists &&
+            (isPathDependent || localSym.exists)) {
+          // Return shortened type for names already in scope
+          val onlyNames = onlyPaths.map(_._1)
+          val removed = removePrefixes(refNoThis, onlyNames)
+          val shortenedAndRenamedType = removed.transform {
```

Comment:
I think that is a good idea. I'm not religious about it, but I think the rule of 30 is nice to keep in mind https://github.com/databricks/scala-style-guide#rule_of_30